File: implementations/javascript/check.js
const fs = require('fs');
const path = require('path');
const { CIDS_DIR, computeCid, downloadCid } = require('./cid');

const baseUrl = 'https://256t.org';

async function main() {
  const mismatches = [];
  const downloadFailures = [];
  let count = 0;

  const entries = fs.readdirSync(CIDS_DIR).sort();
  
  for (const entry of entries) {
    const fullPath = path.join(CIDS_DIR, entry);
    if (fs.statSync(fullPath).isDirectory()) {
      continue;
    }
    count += 1;
    
    const localContent = fs.readFileSync(fullPath);
    const expected = computeCid(localContent);
    
    // Check local CID file
    if (expected !== entry) {
      mismatches.push([entry, expected]);
    }
    
    // Check downloaded content
    try {
      const { content, computed, isValid } = await downloadCid(baseUrl, entry);
      if (!isValid) {
        downloadFailures.push([entry, computed]);
      } else if (!content.equals(localContent)) {
        downloadFailures.push([entry, 'content mismatch with local file']);
      }
    } catch (error) {
      downloadFailures.push([entry, error.message]);
    }
  }

  let hasErrors = false;

  if (mismatches.length > 0) {
    console.log('Found CID mismatches:');
    for (const [entry, expected] of mismatches) {
      console.log(`- ${entry} should be ${expected}`);
    }
    hasErrors = true;
  }

  if (downloadFailures.length > 0) {
    console.error('Found download validation failures:');
    for (const [cid, error] of downloadFailures) {
      console.error(`- ${cid}: ${error}`);
    }
    hasErrors = true;
  }

  if (hasErrors) {
    process.exit(1);
  }

  console.log(`All ${count} CID files match their contents.`);
  console.log(`All ${count} downloaded CIDs are valid.`);
}

main().catch((error) => {
  console.error('Unexpected error:', error);
  process.exit(1);
});

File: implementations/r/check.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(digest)
  library(base64enc)
})

source(file.path(dirname(normalizePath(sub("^--file=", "", grep("^--file=", commandArgs(FALSE), value = TRUE)[1]))), "cid.R"))

read_content <- function(path) {
  size <- file.info(path)$size
  readBin(path, what = "raw", n = size)
}

main <- function() {
  entries <- list.files(CIDS_DIR, full.names = TRUE)
  entries <- entries[!file.info(entries)$isdir]
  entries <- sort(entries)

  mismatches <- list()
  download_failures <- list()
  count <- 0
  base_url <- "https://256t.org"

  for (entry in entries) {
    count <- count + 1
    cid <- basename(entry)
    local_content <- read_content(entry)
    expected <- compute_cid(local_content)
    
    # Check local CID file
    if (!identical(cid, expected)) {
      mismatches[[length(mismatches) + 1]] <- list(cid = cid, expected = expected)
    }
    
    # Check downloaded content
    tryCatch({
      result <- download_cid(base_url, cid)
      if (!result$is_valid) {
        download_failures[[length(download_failures) + 1]] <- list(cid = cid, error = result$computed)
      } else if (!identical(result$content, local_content)) {
        download_failures[[length(download_failures) + 1]] <- list(cid = cid, error = "content mismatch with local file")
      }
    }, error = function(e) {
      download_failures[[length(download_failures) + 1]] <<- list(cid = cid, error = e$message)
    })
  }

  has_errors <- FALSE

  if (length(mismatches) > 0) {
    cat("Found CID mismatches:\n")
    for (mismatch in mismatches) {
      cat(sprintf("- %s should be %s\n", mismatch$cid, mismatch$expected))
    }
    has_errors <- TRUE
  }

  if (length(download_failures) > 0) {
    cat("Found download validation failures:\n", file = stderr())
    for (failure in download_failures) {
      cat(sprintf("- %s: %s\n", failure$cid, failure$error), file = stderr())
    }
    has_errors <- TRUE
  }

  if (has_errors) {
    quit(status = 1)
  }

  cat(sprintf("All %d CID files match their contents.\n", count))
  cat(sprintf("All %d downloaded CIDs are valid.\n", count))
}

main()
